File: clients/php/src/Textmark.php
<?php

declare(strict_types=1);

namespace Sharptown\Client;

/**
 * A text overlay composited onto the result, rendered server-side. Pass it to
 * {@link TransformBuilder::composite()}.
 *
 * @example
 * $st->transform($file)
 *     ->composite((new Textmark('© Acme'))->size(48)->color('white')->rotate(-30)->tile())
 *     ->convert('webp');
 */
final class Textmark implements CompositeMark
{
    /** @var array<string, mixed> */
    private array $spec = ['type' => 'text'];

    public function __construct(string $text)
    {
        $this->spec['text'] = $text;
    }

    /** Font size in pixels. */
    public function size(int $value): self
    {
        $this->spec['size'] = Operations::toPositiveInt($value, 'textmark size');
        return $this;
    }

    /** Text colour (any CSS colour). */
    public function color(string $value): self
    {
        $this->spec['color'] = $value;
        return $this;
    }

    /** Font family. */
    public function font(string $value): self
    {
        $this->spec['font'] = $value;
        return $this;
    }

    /** Font weight (e.g. `bold`). */
    public function weight(string $value): self
    {
        $this->spec['weight'] = $value;
        return $this;
    }

    /** Background colour painted behind the text tile. */
    public function background(string $value): self
    {
        $this->spec['background'] = $value;
        return $this;
    }

    /** Rotates the text by degrees. */
    public function rotate(int $degrees): self
    {
        $this->spec['rotate'] = $degrees;
        return $this;
    }

    /** Text opacity (0–1). */
    public function opacity(int|float $value): self
    {
        $this->spec['opacity'] = Operations::toRange($value, 'textmark opacity', 0, 1);
        return $this;
    }

    /** Placement gravity. */
    public function gravity(string $value): self
    {
        $this->spec['gravity'] = $value;
        return $this;
    }

    /** Places the text at ($x, $y) from the top-left. */
    public function offset(int $x, int $y): self
    {
        $this->spec['x'] = $x;
        $this->spec['y'] = $y;
        return $this;
    }

    /** Repeats the text across the whole image. */
    public function tile(bool $enabled = true): self
    {
        $this->spec['tile'] = $enabled;
        return $this;
    }

    public function resolve(): array
    {
        return ['spec' => $this->spec, 'bytes' => null];
    }
}

File: clients/elixir/lib/sharptown/textmark.ex
defmodule Sharptown.Textmark do
  @moduledoc """
  A text overlay composited onto the result, rendered server-side. Pass it to
  `Sharptown.composite/2`.

  ## Example

      Sharptown.Textmark.new("© Acme")
      |> Sharptown.Textmark.size(48)
      |> Sharptown.Textmark.color("white")
      |> Sharptown.Textmark.tile()
  """

  defstruct spec: %{"type" => "text"}

  @type t :: %__MODULE__{spec: map()}

  @doc "Creates a text watermark."
  def new(text) when is_binary(text), do: %__MODULE__{spec: %{"type" => "text", "text" => text}}

  @doc "Font size in pixels."
  def size(%__MODULE__{} = m, value), do: put(m, "size", value)

  @doc "Text colour (any CSS colour)."
  def color(%__MODULE__{} = m, value), do: put(m, "color", value)

  @doc "Font family."
  def font(%__MODULE__{} = m, value), do: put(m, "font", value)

  @doc "Font weight (e.g. `bold`)."
  def weight(%__MODULE__{} = m, value), do: put(m, "weight", value)

  @doc "Background colour painted behind the text tile."
  def background(%__MODULE__{} = m, value), do: put(m, "background", value)

  @doc "Rotates the text by degrees."
  def rotate(%__MODULE__{} = m, degrees), do: put(m, "rotate", degrees)

  @doc "Text opacity (0–1)."
  def opacity(%__MODULE__{} = m, value), do: put(m, "opacity", value)

  @doc "Placement gravity."
  def gravity(%__MODULE__{} = m, value), do: put(m, "gravity", value)

  @doc "Places the text at `(x, y)` from the top-left."
  def offset(%__MODULE__{} = m, x, y), do: m |> put("x", x) |> put("y", y)

  @doc "Repeats the text across the whole image."
  def tile(%__MODULE__{} = m, enabled \\ true), do: put(m, "tile", enabled)

  @doc false
  def resolve(%__MODULE__{spec: spec}), do: {spec, nil}

  defp put(%__MODULE__{spec: spec} = m, key, value), do: %{m | spec: Map.put(spec, key, value)}
end
